### DISCUSSION TITLE
Are there any performance implications to the spacing of entity IDs?

### DISCUSSION THREAD

**Discussion by Net5F** [skypjack/entt#1050] [Q&A] [answered]
Hello, I'm trying to decide whether I should persist my entities with their `entt::entity` ID directly, or if I should give each entity a separate, persistent ID. The difference with a separate ID would be that, on startup, I could just use `create()`. If I persisted the `entt::entity` directly, I would need to use `create(hint)` and their IDs would be more sparse.

Are there any situations where the closeness of IDs has significant performance implications?

**Comment by skypjack** [comment, marked as answer]
There is a subtle difference actually to make it possible to recycle identifiers without having to keep track of all _holes_ in your set.
When generating identifier N, it triggers the generation of all non-existing identifiers with an entity id lower than its own, if any exist.
It's not a big deal since you're going to generate them all anyway sooner or later. Therefore, I don't think it will cause you any problems.

**Reply by Net5F** [reply]
Perfect, thanks for the explanation 😄

**Reply by skypjack** [reply]
You're welcome.

**Reply by Net5F** [reply]
Funny timing, I just ran into a related question. If you create an entity as such: `registry.create(static_cast<entt::entity>(1))`, is `registry.valid(static_cast<entt::entity>(0))` afterwards meant to return `true`? That's what I appear to be seeing, and I realized this is probably what you were explaining.

My game is networked, so I have a server creating entities locally and telling clients to replicate them. If one of my clients is told to create entity 20 followed by entity 15, it will do a check for if 15 already exists, then fail since it returns true. I supposed I could skip the `registry.valid()` check and just `registry.create()` and see if the returned ID matches the desired ID. Does that sound reasonable?

**Reply by skypjack** [reply]
In fact, it depends. If entity 0 exists already and got destroyed before you create entity 1, then the call to `valid` isn't guaranteed to return true.
If you do that on an empty registry instead, then yes, it returns true. In order to create entity 1, the registry also creates entity 0 and puts it in the list of _available entities_ (the ones to return if you invoke `create()` without arguments). Therefore, technically speaking the identifier exists.
What you want to check is probably that is exists and that its index `as in `index(entity)` is lower than `in_use()`. What I want probably is a new `in_use(entity)` function instead in order to avoid this ambiguity. 🙂

**Reply by Net5F** [reply]
Ah, that makes sense. Thanks again for the explanation 😄
